File: VB/DevExpress.Xpf.NavBar.Extensions/NavBarMVVMAttachedBehavior.vb
' Developer Express Code Central Example:
' How to use the DXNavBar in a MVVM application
' 
' This example demonstrates how to use the DXNavBar with MVVM patter using special
' attached behavior.
' See http://www.devexpress.com/scid=K18540 for more
' information.
' 
' You can find sample updates and versions for different programming languages here:
' http://www.devexpress.com/example=E3521

Imports System.Windows.Interactivity
Imports System.Windows
Imports System.Collections
Imports System.Collections.Specialized
Imports System
Imports System.Windows.Data
Imports DevExpress.Xpf.NavBar
Namespace NavBarExtensions
    Public Class NavBarMVVMAttachedBehavior
        Inherits Behavior(Of NavBarControl)

        #Region "DP"
        Public Shared ReadOnly ItemsSourceProperty As DependencyProperty = DependencyProperty.Register("ItemsSource", GetType(IEnumerable), GetType(NavBarMVVMAttachedBehavior), New PropertyMetadata(Nothing, Sub(d,e) CType(d, NavBarMVVMAttachedBehavior).OnItemsSourceChanged(DirectCast(e.OldValue, IEnumerable))))
        Public Shared ReadOnly GroupStyleProperty As DependencyProperty = DependencyProperty.Register("GroupStyle", GetType(Style), GetType(NavBarMVVMAttachedBehavior), New PropertyMetadata(Nothing))
        Public Shared ReadOnly ItemStyleProperty As DependencyProperty = DependencyProperty.Register("ItemStyle", GetType(Style), GetType(NavBarMVVMAttachedBehavior), New PropertyMetadata(Nothing))

        Public Shared ReadOnly NavBarGroupStylerProperty As DependencyProperty = DependencyProperty.RegisterAttached("NavBarGroupStyler", GetType(NavBarGroupWrapper), GetType(NavBarMVVMAttachedBehavior), New PropertyMetadata(Nothing))
        Public Shared ReadOnly NavBarItemStylerProperty As DependencyProperty = DependencyProperty.RegisterAttached("NavBarItemStyler", GetType(NavBarItemWrapper), GetType(NavBarMVVMAttachedBehavior), New PropertyMetadata(Nothing))
        Public Shared Function GetNavBarGroupStyler(ByVal obj As NavBarGroup) As NavBarGroupWrapper
            Return CType(obj.GetValue(NavBarGroupStylerProperty), NavBarGroupWrapper)
        End Function
        Public Shared Sub SetNavBarGroupStyler(ByVal obj As NavBarGroup, ByVal value As NavBarGroupWrapper)
            obj.SetValue(NavBarGroupStylerProperty, value)
        End Sub
        Public Shared Function GetNavBarItemStyler(ByVal obj As NavBarItem) As NavBarItemWrapper
            Return CType(obj.GetValue(NavBarItemStylerProperty), NavBarItemWrapper)
        End Function
        Public Shared Sub SetNavBarItemStyler(ByVal obj As NavBarItem, ByVal value As NavBarItemWrapper)
            obj.SetValue(NavBarItemStylerProperty, value)
        End Sub

        Public Property ItemsSource() As IEnumerable
            Get
                Return DirectCast(GetValue(ItemsSourceProperty), IEnumerable)
            End Get
            Set(ByVal value As IEnumerable)
                SetValue(ItemsSourceProperty, value)
            End Set
        End Property
        Public Property GroupStyle() As Style
            Get
                Return CType(GetValue(GroupStyleProperty), Style)
            End Get
            Set(ByVal value As Style)
                SetValue(GroupStyleProperty, value)
            End Set
        End Property
        Public Property ItemStyle() As Style
            Get
                Return CType(GetValue(ItemStyleProperty), Style)
            End Get
            Set(ByVal value As Style)
                SetValue(ItemStyleProperty, value)
            End Set
        End Property
        #End Region
        Public Function GetNavBarGroupContainerFromItem(ByVal data As Object) As NavBarGroup
            If Not IsInitialized Then
                Return Nothing
            End If
            For Each gr As NavBarGroup In NavBar.Groups
                If gr.DataContext Is data Then
                    Return gr
                End If
            Next gr
            Return Nothing
        End Function
        Public Function GetItemFromNavBarGroupContainer(ByVal gr As NavBarGroup) As Object
            Return gr.DataContext
        End Function

        Protected ReadOnly Property NavBar() As NavBarControl
            Get
                Return AssociatedObject
            End Get
        End Property
        Private privateIsInitialized As Boolean
        Protected Property IsInitialized() As Boolean
            Get
                Return privateIsInitialized
            End Get
            Private Set(ByVal value As Boolean)
                privateIsInitialized = value
            End Set
        End Property

        Protected Overrides Sub OnAttached()
            MyBase.OnAttached()
            If Not IsInitialized Then
                TryInitialize()
                Return
            End If
        End Sub
        Protected Overrides Sub OnDetaching()
            MyBase.OnDetaching()
            If TypeOf ItemsSource Is INotifyCollectionChanged Then
                RemoveHandler DirectCast(ItemsSource, INotifyCollectionChanged).CollectionChanged, AddressOf OnItemsSourceCollectionChanged
            End If
            ClearNavBarGroups()
            IsInitialized = False
        End Sub
        Protected Overridable Sub OnItemsSourceChanged(ByVal oldSource As IEnumerable)
            If Not IsInitialized Then
                TryInitialize()
                Return
            End If
            If TypeOf oldSource Is INotifyCollectionChanged Then
                RemoveHandler DirectCast(oldSource, INotifyCollectionChanged).CollectionChanged, AddressOf OnItemsSourceCollectionChanged
            End If
            ClearNavBarGroups()
            If ItemsSource Is Nothing Then
                IsInitialized = False
                Return
            End If
            For Each obj As Object In ItemsSource
                AddNavBarGroup(obj)
            Next obj
            If TypeOf ItemsSource Is INotifyCollectionChanged Then
                AddHandler DirectCast(ItemsSource, INotifyCollectionChanged).CollectionChanged, AddressOf OnItemsSourceCollectionChanged
            End If
        End Sub
        Private Sub TryInitialize()
            If IsInitialized Then
                Return
            End If
            If NavBar Is Nothing OrElse ItemsSource Is Nothing Then
                Return
            End If
            IsInitialized = True
            For Each obj As Object In ItemsSource
                AddNavBarGroup(obj)
            Next obj
            If TypeOf ItemsSource Is INotifyCollectionChanged Then
                AddHandler DirectCast(ItemsSource, INotifyCollectionChanged).CollectionChanged, AddressOf OnItemsSourceCollectionChanged
            End If
        End Sub
        Private Sub OnItemsSourceCollectionChanged(ByVal sender As Object, ByVal e As NotifyCollectionChangedEventArgs)
            If e.Action = NotifyCollectionChangedAction.Add AndAlso e.NewItems IsNot Nothing Then
                For i As Integer = 0 To e.NewItems.Count - 1
                    If e.NewStartingIndex = -1 Then
                        AddNavBarGroup(e.NewItems(i))
                    Else
                        AddNavBarGroup(e.NewItems(i), i + e.NewStartingIndex)
                    End If
                Next i
            End If
            If e.Action = NotifyCollectionChangedAction.Remove AndAlso e.OldItems IsNot Nothing Then
                For i As Integer = 0 To e.OldItems.Count - 1
                    RemoveNavBarGroup(e.OldItems(i))
                Next i
            End If
            If e.Action = NotifyCollectionChangedAction.Replace Then
                If e.OldItems IsNot Nothing Then
                    For i As Integer = 0 To e.OldItems.Count - 1
                        RemoveNavBarGroup(e.OldItems(i))
                    Next i
                End If
                If e.NewItems IsNot Nothing Then
                    For i As Integer = 0 To e.NewItems.Count - 1
                        If e.NewStartingIndex = -1 Then
                            AddNavBarGroup(e.NewItems(i))
                        Else
                            AddNavBarGroup(e.NewItems(i), i + e.NewStartingIndex)
                        End If
                    Next i
                End If
            End If
            If e.Action = NotifyCollectionChangedAction.Reset Then
                ClearNavBarGroups()
            End If
        End Sub

        Private Sub ClearNavBarGroups()
            If Not IsInitialized Then
                Return
            End If
            For Each gr As NavBarGroup In NavBar.Groups
                RemoveNavBarGroup(gr)
            Next gr
        End Sub
        Private Sub AddNavBarGroup(ByVal obj As Object)
            If Not IsInitialized Then
                Return
            End If
            NavBar.Groups.Add(CreateNavBarGroup(obj))
        End Sub
        Private Sub AddNavBarGroup(ByVal obj As Object, ByVal index As Integer)
            If Not IsInitialized Then
                Return
            End If
            NavBar.Groups.Insert(index, CreateNavBarGroup(obj))
        End Sub
        Private Sub RemoveNavBarGroup(ByVal obj As Object)
            If Not IsInitialized Then
                Return
            End If
            RemoveNavBarGroup(GetNavBarGroupContainerFromItem(obj))
        End Sub
        Private Sub RemoveNavBarGroup(ByVal gr As NavBarGroup)
            Dim styler As NavBarGroupWrapper = NavBarMVVMAttachedBehavior.GetNavBarGroupStyler(gr)
            DirectCast(styler, IDisposable).Dispose()
            RemoveHandler styler.NavBarItemsCollectionChanged, AddressOf OnStylerNavBarItemsCollectionChanged
        End Sub
        Private Function CreateNavBarGroup(ByVal obj As Object) As NavBarGroup
            Dim gr As New NavBarGroup() With {.DataContext = obj}
            Dim styler As New NavBarGroupWrapper(GroupStyle, gr, obj)
            AddHandler styler.NavBarItemsCollectionChanged, AddressOf OnStylerNavBarItemsCollectionChanged
            styler.Init()
            Return gr
        End Function

        Protected Overridable Sub OnGroupItemAdding(ByVal item As NavBarItem, ByVal dataObject As Object)
            Dim styler As New NavBarItemWrapper(ItemStyle, item, dataObject)
            styler.Init()
        End Sub
        Protected Overridable Sub OnGroupItemRemoving(ByVal item As NavBarItem, ByVal dataObject As Object)
            Dim styler As NavBarItemWrapper = NavBarMVVMAttachedBehavior.GetNavBarItemStyler(item)
            DirectCast(styler, IDisposable).Dispose()
        End Sub

        Private Sub OnStylerNavBarItemsCollectionChanged(ByVal groupStyler As NavBarGroupWrapper, ByVal e As NotifyCollectionChangedEventArgs, ByVal isDisposing As Boolean)
            If e.Action = NotifyCollectionChangedAction.Add AndAlso e.NewItems IsNot Nothing Then
                For Each item As NavBarItem In e.NewItems
                    OnGroupItemAdding(item, groupStyler.GetNavBarItemDataObject(item))
                Next item
            End If
            If e.Action = NotifyCollectionChangedAction.Remove AndAlso e.OldItems IsNot Nothing Then
                For Each item As NavBarItem In e.OldItems
                    OnGroupItemRemoving(item, groupStyler.GetNavBarItemDataObject(item))
                Next item
            End If
            If e.Action = NotifyCollectionChangedAction.Replace Then
                If e.OldItems IsNot Nothing Then
                    For Each item As NavBarItem In e.OldItems
                        OnGroupItemRemoving(item, groupStyler.GetNavBarItemDataObject(item))
                    Next item
                End If
                If e.NewItems IsNot Nothing Then
                    For Each item As NavBarItem In e.NewItems
                        OnGroupItemAdding(item, groupStyler.GetNavBarItemDataObject(item))
                    Next item
                End If
            End If
            If e.Action = NotifyCollectionChangedAction.Reset Then
                For Each item As NavBarItem In groupStyler.NavBarItems
                    OnGroupItemRemoving(item, groupStyler.GetNavBarItemDataObject(item))
                Next item
                If Not isDisposing Then
                    For Each item As NavBarItem In groupStyler.NavBarItems
                        OnGroupItemAdding(item, groupStyler.GetNavBarItemDataObject(item))
                    Next item
                End If
            End If
        End Sub
    End Class
End Namespace

File: CS/DevExpress.Xpf.NavBar.Extensions/NavBarMVVMAttachedBehavior.cs
// Developer Express Code Central Example:
// How to use the DXNavBar in a MVVM application
// 
// This example demonstrates how to use the DXNavBar with MVVM patter using special
// attached behavior.
// See http://www.devexpress.com/scid=K18540 for more
// information.
// 
// You can find sample updates and versions for different programming languages here:
// http://www.devexpress.com/example=E3521

using System.Windows.Interactivity;
using System.Windows;
using System.Collections;
using System.Collections.Specialized;
using System;
using System.Windows.Data;
using DevExpress.Xpf.NavBar;
namespace NavBarExtensions {
    public class NavBarMVVMAttachedBehavior : Behavior<NavBarControl> {
        #region DP
        public static readonly DependencyProperty ItemsSourceProperty =
            DependencyProperty.Register("ItemsSource", typeof(IEnumerable), typeof(NavBarMVVMAttachedBehavior),
            new PropertyMetadata(null, (d,e) => ((NavBarMVVMAttachedBehavior)d).OnItemsSourceChanged((IEnumerable)e.OldValue)));
        public static readonly DependencyProperty GroupStyleProperty =
            DependencyProperty.Register("GroupStyle", typeof(Style), typeof(NavBarMVVMAttachedBehavior), new PropertyMetadata(null));
        public static readonly DependencyProperty ItemStyleProperty =
            DependencyProperty.Register("ItemStyle", typeof(Style), typeof(NavBarMVVMAttachedBehavior), new PropertyMetadata(null));

        public static readonly DependencyProperty NavBarGroupStylerProperty =
            DependencyProperty.RegisterAttached("NavBarGroupStyler", typeof(NavBarGroupWrapper), typeof(NavBarMVVMAttachedBehavior), new PropertyMetadata(null));
        public static readonly DependencyProperty NavBarItemStylerProperty =
            DependencyProperty.RegisterAttached("NavBarItemStyler", typeof(NavBarItemWrapper), typeof(NavBarMVVMAttachedBehavior), new PropertyMetadata(null));
        public static NavBarGroupWrapper GetNavBarGroupStyler(NavBarGroup obj) {
            return (NavBarGroupWrapper)obj.GetValue(NavBarGroupStylerProperty);
        }
        public static void SetNavBarGroupStyler(NavBarGroup obj, NavBarGroupWrapper value) {
            obj.SetValue(NavBarGroupStylerProperty, value);
        }
        public static NavBarItemWrapper GetNavBarItemStyler(NavBarItem obj) {
            return (NavBarItemWrapper)obj.GetValue(NavBarItemStylerProperty);
        }
        public static void SetNavBarItemStyler(NavBarItem obj, NavBarItemWrapper value) {
            obj.SetValue(NavBarItemStylerProperty, value);
        }

        public IEnumerable ItemsSource {
            get { return (IEnumerable)GetValue(ItemsSourceProperty); }
            set { SetValue(ItemsSourceProperty, value); }
        }
        public Style GroupStyle {
            get { return (Style)GetValue(GroupStyleProperty); }
            set { SetValue(GroupStyleProperty, value); }
        }
        public Style ItemStyle {
            get { return (Style)GetValue(ItemStyleProperty); }
            set { SetValue(ItemStyleProperty, value); }
        }
        #endregion
        public NavBarGroup GetNavBarGroupContainerFromItem(object data) {
            if(!IsInitialized) return null;
            foreach(NavBarGroup gr in NavBar.Groups)
                if(gr.DataContext == data) return gr;
            return null;
        }
        public object GetItemFromNavBarGroupContainer(NavBarGroup gr) {
            return gr.DataContext;
        }

        protected NavBarControl NavBar { get { return AssociatedObject; } }
        protected bool IsInitialized { get; private set; }
        
        protected override void OnAttached() {
            base.OnAttached();
            if(!IsInitialized) {
                TryInitialize();
                return;
            }
        }
        protected override void OnDetaching() {
            base.OnDetaching();
            if(ItemsSource is INotifyCollectionChanged) ((INotifyCollectionChanged)ItemsSource).CollectionChanged -= OnItemsSourceCollectionChanged;
            ClearNavBarGroups();
            IsInitialized = false;
        }
        protected virtual void OnItemsSourceChanged(IEnumerable oldSource) {
            if(!IsInitialized) {
                TryInitialize();
                return;
            }
            if(oldSource is INotifyCollectionChanged) ((INotifyCollectionChanged)oldSource).CollectionChanged -= OnItemsSourceCollectionChanged;
            ClearNavBarGroups();
            if(ItemsSource == null) {
                IsInitialized = false;
                return;
            }
            foreach(object obj in ItemsSource) AddNavBarGroup(obj);
            if(ItemsSource is INotifyCollectionChanged) ((INotifyCollectionChanged)ItemsSource).CollectionChanged += OnItemsSourceCollectionChanged;
        }
        void TryInitialize() {
            if(IsInitialized) return;
            if(NavBar == null || ItemsSource == null) return;
            IsInitialized = true;
            foreach(object obj in ItemsSource) AddNavBarGroup(obj);
            if(ItemsSource is INotifyCollectionChanged) ((INotifyCollectionChanged)ItemsSource).CollectionChanged += OnItemsSourceCollectionChanged;
        }
        void OnItemsSourceCollectionChanged(object sender, NotifyCollectionChangedEventArgs e) {
            if(e.Action == NotifyCollectionChangedAction.Add && e.NewItems != null) {
                for(int i = 0; i < e.NewItems.Count; i++)
                    if(e.NewStartingIndex == -1) AddNavBarGroup(e.NewItems[i]);
                    else AddNavBarGroup(e.NewItems[i], i + e.NewStartingIndex);
            }
            if(e.Action == NotifyCollectionChangedAction.Remove && e.OldItems != null) {
                for(int i = 0; i < e.OldItems.Count; i++) RemoveNavBarGroup(e.OldItems[i]);
            }
            if(e.Action == NotifyCollectionChangedAction.Replace) {
                if(e.OldItems != null)
                    for(int i = 0; i < e.OldItems.Count; i++) RemoveNavBarGroup(e.OldItems[i]);
                if(e.NewItems != null)
                    for(int i = 0; i < e.NewItems.Count; i++) {
                        if(e.NewStartingIndex == -1) AddNavBarGroup(e.NewItems[i]);
                        else AddNavBarGroup(e.NewItems[i], i + e.NewStartingIndex);
                    }
            }
            if(e.Action == NotifyCollectionChangedAction.Reset)
                ClearNavBarGroups();
        }

        void ClearNavBarGroups() {
            if(!IsInitialized) return;
            foreach(NavBarGroup gr in NavBar.Groups)
                RemoveNavBarGroup(gr);
        }
        void AddNavBarGroup(object obj) {
            if(!IsInitialized) return;
            NavBar.Groups.Add(CreateNavBarGroup(obj));
        }
        void AddNavBarGroup(object obj, int index) {
            if(!IsInitialized) return;
            NavBar.Groups.Insert(index, CreateNavBarGroup(obj));
        }
        void RemoveNavBarGroup(object obj) {
            if(!IsInitialized) return;
            RemoveNavBarGroup(GetNavBarGroupContainerFromItem(obj));
        }
        void RemoveNavBarGroup(NavBarGroup gr) {
            NavBarGroupWrapper styler = NavBarMVVMAttachedBehavior.GetNavBarGroupStyler(gr);
            ((IDisposable)styler).Dispose();
            styler.NavBarItemsCollectionChanged -= OnStylerNavBarItemsCollectionChanged;
        }
        NavBarGroup CreateNavBarGroup(object obj) {
            NavBarGroup gr = new NavBarGroup() { DataContext = obj };
            NavBarGroupWrapper styler = new NavBarGroupWrapper(GroupStyle, gr, obj);
            styler.NavBarItemsCollectionChanged += OnStylerNavBarItemsCollectionChanged;
            styler.Init();
            return gr;
        }

        protected virtual void OnGroupItemAdding(NavBarItem item, object dataObject) {
            NavBarItemWrapper styler = new NavBarItemWrapper(ItemStyle, item, dataObject);
            styler.Init();
        }
        protected virtual void OnGroupItemRemoving(NavBarItem item, object dataObject) {
            NavBarItemWrapper styler = NavBarMVVMAttachedBehavior.GetNavBarItemStyler(item);
            ((IDisposable)styler).Dispose();
        }
        
        void OnStylerNavBarItemsCollectionChanged(NavBarGroupWrapper groupStyler, NotifyCollectionChangedEventArgs e, bool isDisposing) {
            if(e.Action == NotifyCollectionChangedAction.Add && e.NewItems != null) {
                foreach(NavBarItem item in e.NewItems)
                    OnGroupItemAdding(item, groupStyler.GetNavBarItemDataObject(item));
            }
            if(e.Action == NotifyCollectionChangedAction.Remove && e.OldItems != null) {
                foreach(NavBarItem item in e.OldItems)
                    OnGroupItemRemoving(item, groupStyler.GetNavBarItemDataObject(item));
            }
            if(e.Action == NotifyCollectionChangedAction.Replace) {
                if(e.OldItems != null) {
                    foreach(NavBarItem item in e.OldItems)
                        OnGroupItemRemoving(item, groupStyler.GetNavBarItemDataObject(item));
                }
                if(e.NewItems != null) {
                    foreach(NavBarItem item in e.NewItems)
                        OnGroupItemAdding(item, groupStyler.GetNavBarItemDataObject(item));
                }
            }
            if(e.Action == NotifyCollectionChangedAction.Reset) {
                foreach(NavBarItem item in groupStyler.NavBarItems)
                    OnGroupItemRemoving(item, groupStyler.GetNavBarItemDataObject(item));
                if(!isDisposing) {
                    foreach(NavBarItem item in groupStyler.NavBarItems)
                        OnGroupItemAdding(item, groupStyler.GetNavBarItemDataObject(item));
                }
            }
        }
    }
}